src/server/pubsub/data/process.c

```diff
@@ -0,0 +1,337 @@
+
+static bool process_init = false;
+static process_metrics_st *process_metrics = NULL;
+
+
+/* "publish topic msg"
+ * reply: # listeners subscribed
+ */
+static void
+command_publish(struct response *rsp, struct request *req, struct buf_sock *s)
+{
+    struct element *el_t, *el_m; /* topic & message */
+    struct element el_r = {.type = ELEM_INT}; /* reply, an integer */
+    struct listener *l;
+    struct topic *t;
+    uint32_t nsub = 0;
+
+    el_t = array_get(req->token, 1);
+    if (el_t->type != ELEM_BULK) {
+        /* handle error */
+    };
+    el_m = array_get(req->token, 2);
+    if (el_m->type != ELEM_BULK) {
+        /* handle error */
+    };
+
+    log_verb("publish from buf_sock %p", s);
+
+    l = listener_get(s);
+    if (l != NULL) {
+        log_error("found listener at %p: subscriber cannot publish", l);
+        /* handle error */
+    }
+
+    t = topic_get(&el_t->bstr);
+    if (t == NULL) {
+        log_verb("no listener on topic %.*s, ignore", el_t->bstr.len,
+                el_t->bstr.data);
+    } else { /* copy message to listener's buffer, this should be optimized later */
+        struct index_node *node;
+
+        nsub = t->nsub;
+        TAILQ_FOREACH(node, t->idx, i_tqe) {
+            l = (struct listener *)node->obj;
+            struct buf_sock *ss = l->s;
+            compose_array_header(&ss->wbuf, 3);
+            compose_element(&ss->wbuf, &el_message);
+            compose_element(&ss->wbuf, el_t);
+            compose_element(&ss->wbuf, el_m);
+            /* register listener's fd for write event */
+            event_add_write(ctx->evb, ss->hdl->wid(ss->ch), ss);
+        }
+    }
+
+    el_r.num = nsub;
+    compose_element(&s->wbuf, &el_r);
+}
+
+/* "subscribe topic [topic ...]"
+ * no reply
+ */
+static void
+command_subscribe(struct response *rsp, struct request *req, struct buf_sock *s)
+{
+    struct element *el;
+    struct element el_s = {.type = ELEM_INT}; /* # topics subscribed, an integer */
+    struct listener *l;
+    struct topic *t;
+    uint32_t ntopic = req->token->nelem - 1;
+
+    log_verb("subscribe buf_sock %p from topics", s);
+
+    l = listener_get(s);
+    if (l == NULL) {
+        log_verb("create new listener for %p", s);
+        l = listener_create(s);
+        listener_put(l);
+    }
+
+    for (int i = 1; i <= ntopic; i++) {
+        el = array_get(req->token, i);
+        if (el->type != ELEM_BULK) {
+            /* handle error */
+        };
+
+        t = topic_get(&el->bstr);
+        if (t == NULL) {
+            log_verb("creating topic %.*s", el->bstr.len, el->bstr.data);
+            t = topic_create(&el->bstr);
+            /* handle error */
+            topic_put(t);
+        }
+
+        log_verb("subscribing to topic %.*s", el->bstr.len, el->bstr.data);
+        if (!topic_add_listener(t, l)) {
+            log_debug("listener not added");
+        }
+        if (!listener_add_topic(l, t)) {
+            log_debug("topic not added");
```

Comment:
Let me keep you comment in mind for the moment- I need to handle errors in a number of other places as well but it's not immediately clear what Redis does/should do in this case. I'll research more and come back to this.